---
title: "Final Presentation"
author: "Ian Mc Farlane, Bram Stults"
date: "2025-04-16"
output: slidy_presentation
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

## Introduction

Meta-omic analysis of RNA-sequencing data investigating CVID across tissue samples

**Stakeholder**

-   Worked with W. Evan Johnson of Rutgers University

**Mission**

-   Deployed bioinformatic models to analyze genetic count data

-   Focussed on CVID: a clinically defined disease, but which is molecularly not well-defined

-   Organized a complete code-base to support immunological research and publishability

## Context

This analysis is subsequent to work of Dr. Paul Maglione : '*Convergence of cytokine dysregulation and antibody deficiency in common variable immunodeficiency with inflammatory complications*' :

-   Preceding study involved LPS stimulation of peripheral blood tissue and identified "potential significance of reduced LPS-specific antibodies"... "suggesting a a 2-hit model of pathogenesis requiring further exploration"

## Local and Global Impact

We hope our analysis will contribute as support system and a viewpoint on :

**Locally**

-   Building a code repository that is useful both as a tool and evidence supporting publishable results
-   Supporting research of Drs. W. Evan Johnson of Rutgers and P.J. Maglione of Boston U.

**Globally**

-   Expanding understanding of less understood immunological topic, contributing to body of evidence

-   Supporting the immunological research community and hopefully providing a data point one day improving care

## Continuing Professional Development

*More than 5 topics outside of the curriculum were discussed or needed for project completion:*

-   Jaccard Coefficient
-   **Permutation Tests** for analysis design validation
-   Bioinformatics pipelines for Differential Expression Analysis and Pathway Analysis *(Summarized Experiment, DESeq2, Limma, and enrichR Libraries)*
-   Introductory Immunology and Clinical Practices
-   Professional Journal Writing

## Data Science Lifecycle

1.  Data acquisition and representativeness
2.  Data management
3.  Data preparation and integration
4.  Data analysis
5.  Model development and deployment
6.  Visualization and communication of the knowledge obtained from the data

## 1. Data acquisition and representativeness

Dr Johnson provided 3 separate data sets directly. These were developed or measured by research teams at Boston University and Rutgers:

1.  Nasopharynx sampling of controls and CVID-diagnosed patients
2.  Peripheral Blood Mononuclear Cell samples (white blood cells found in bloodstream) for controls and CVID
3.  Nasal and Blood samples from patients either control or living with Tuberculosis

**As with other genomic studies, this analysis was subject to problem of "small N big P"**

## 2. Data management

-   We learned and used the Summarized Experiment (S4) data structure

-   Patient anonymity was maintained throughout

-   Maintaining separate databases while sharing versions through Git

## 3. Data preparation and integration

-   Moderate EDA performed to identify and label appropriate treatment groups

-   S4 Assays : simple list-like objects used

-   Use of online pathway libraries (*Ian: please give some technical elaboration /correct / or remove)*

## 4. Data analysis

-   Dimensionality reduction by PCA (axial) and UMAP (maintaining distances)

-   DESeq2 model deployed, to estimate variance and account for overdispersion: gene $i$ in sample $j$ ($K_{ij}$) where $\alpha_i$ is dispersion parameter:

$$
K_{ij} \thicksim NB(\mu_{ij}, \alpha_i)
$$

-   Pathway analysis (*Ian: please elaborate)*

-   Comparative analysis using **Jaccard coefficient** statistic and **permutation tests**

## 5. Model development/deployment

-   Bioinformatic models with broad, longstanding use

-   ...

-   Adjusted analysis for treatment group masking, eventually LPS-stimulation became superfluous

-   ...

## 6. Communication of Knowledge Obtained from Data

-   Creation of subject appropriate statistical plots : UMAP plots, Heatmaps, Volcano plots

-   Adjusting figures for publication-readiness

-   Rslidy for clear and concise presentation, transmission

-   Journal-appropriate writing, in particular for *Methods* section

## Non-technical Discussion

*More than one of legal, professional, ethical, security, or social issues were discussed in depth:*

1.  Samples are described with metadata for statistical analysis, but must be separately analyzed for discretion around patient anonymity
2.  The social **ethics of designations and sensitive usage** were discussed: Ugandan data included **patients living with HIV**. Diagnosis of Tuberculosis may warrant similar sensitivities.
3.  Presently **evolving landscape of clinical research and public funding** were discussed with Dr. Johnson, including professional practices and current risks

## Development Methodology

***Methodology***

-   Worked: \# *I don't know what to say... Any thoughts Ian ?*

-   Didn't Work: inconsistent or shifting coding syntax \# *Please add something...*

***Teamwork***

-   Worked: multi-channel access and version control; different skill-sets; largely clear task division

-   Didn't Work: occasional time conflicts from multiple schedules; skill-set differentials

## Outside Classes

What math and Data Science classes were used in the project?

::::: columns
::: column
**Math Courses**

-   MATH 3700: General Statistical Knowledge
-   MATH 2170: Matrix Mastery and Introduction to Dimension Reduction
-   MATH 3150: Introduction to linear models and advanced statistics
-   MATH 3190: Advanced R and Rmd
:::

::: column
**Support Courses**

-   ANLY 4100: Data Visualization
-   ANLY 4110: Advanced Data Visualization
-   CSCY 2400: Ethical Data Issues
-   CS 2420: Data Structures Basics
-   BIOL 3060: Genetics background
:::
:::::

## Conclusion

## Thank you
