---
title: "Analysis of Immune Deficiency RNA-Seq Data"
author: Ian Mc Farlane, Bram Stults advised by W. Evan Johnson, Ph.D.
date: "`r Sys.Date()`"
header-includes:
   - \usepackage{amsmath}
output:
  html_document:
    code_folding: show
    toc: true
    toc_float: true
    theme: "flatly"
editor_options: 
  chunk_output_type: console
tables: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# load required libraries
library(Rsubread)
library(Rsamtools)
library(SummarizedExperiment)
library(edgeR)
library(DESeq2)
library(ComplexHeatmap)
library(TBSignatureProfiler)
library(umap)
library(tidyverse)
library(DT)
library(GSVA)
library(enrichR)
library(biomaRt)
library(rprojroot)
library(fs)
library(dplyr)

# define pathway libraries
dbs <- c("WikiPathways_2016","KEGG_2016","Panther_2016","Reactome_2016")
```

### Read in dataset

```{r}

# read RNA expression counts and patient data
load(file="data/readMatrix_PBMC_BulkRNAseq_CVID.Rdata")

if (!file.exists("translated_genes.txt")) {
    source("gene_translation.R")
}


mt.counts.df <- read.csv("translated_genes.txt", row.names = 1)
groups <- sample.info$disease 

groups <- ifelse(groups %in% c("UCVID", "CVID"), 
                "Deficient", 
                as.character(groups)) 

# store index of NCVID subjects
remove_NCVID_index <- which(groups == "NCVID")

# update group label vector
groups <- groups[groups != "NCVID"] 
groups <- factor(groups)

# check modified vector
print(groups) 

#updating dataframes
mt.counts.df <- mt.counts.df[,-remove_NCVID_index]
sample.info <- sample.info[-remove_NCVID_index,]
sample.info$disease <- groups

#  re-store results
counts <- mt.counts.df
meta_data <- sample.info

# isolate group indicator variable
group <- meta_data$disease

# aggregate patient and expression data into a single object
se_cvid <- SummarizedExperiment(assays=list(counts=counts),
                     colData = meta_data)


# isolating treatment group experiments
treatments <- meta_data$treatment
ctrl_index <- which(treatments == "ctrl")
LPS_index <- which(treatments == "LPS")

counts_ctrl <- counts[,-LPS_index]
counts_LPS <- counts[,-ctrl_index]

meta_data_ctrl <- meta_data[-LPS_index,]
meta_data_LPS <- meta_data[-ctrl_index,]

se_cvid_ctrl <- SummarizedExperiment(assays=list(counts=counts_ctrl),
                     colData = meta_data_ctrl)
se_cvid_LPS <- SummarizedExperiment(assays=list(counts=counts_LPS),
                     colData = meta_data_LPS)

# create counts per million, log counts and log counts per million features
# and save them into new assays
se_cvid <- mkAssay(se_cvid, log = TRUE, counts_to_CPM = TRUE)
# same for control only
se_cvid_ctrl <- mkAssay(se_cvid_ctrl, log = TRUE, counts_to_CPM = TRUE)
# same for LPS treatment group only
se_cvid_LPS <- mkAssay(se_cvid_LPS, log = TRUE, counts_to_CPM = TRUE)


# display all assays, control and LPS treatment
assays(se_cvid)

# display all assays, control only
assays(se_cvid_ctrl)

# display all assays, LPS treatment only
assays(se_cvid_LPS)

```

### Visualization and Dimension reduction {.tabset}

#### PCA

##### Entire sample - Healthy Control & LPS Treatment

```{r}
# fit PCA model to log cpm data
# note: transpose data to make it tidy
set.seed(1)
pca_out <- prcomp(t(assay(se_cvid,"log_counts_cpm")))

# define dataframe with PCA components and immune deficiency disease (IDC)
pca_plot <- as.data.frame(pca_out$x) # X, Y, ...
pca_plot$disease <- as.factor(se_cvid$disease) # color/group

# plot PC1 and PC2 by IDC
g <- pca_plot %>% ggplot(aes(x=PC1, y=PC2, color=disease)) +
  geom_point(size=1.5) + xlab("PC 1") + ylab("PC 2") +
  geom_text(label=colnames(se_cvid), nudge_y = 5,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("PCA Plot - Entire Sample")

plot(g)
```

#### UMAP

##### Entire sample - Healthy Control & LPS Treatment

```{r}
# fit UMAP model to log cpm data
# note: transpose data to make it tidy
set.seed(1)
umap.defaults$n_neighbors=5
umap_out <- umap(t(assay(se_cvid,"log_counts_cpm")), config=umap.defaults)

# define dataframe with UMAP dimensions and IDC
umap_plot <- as.data.frame(umap_out$layout) # X, Y, ...
umap_plot$disease <- as.factor(se_cvid$disease) # color / group

# plot UMAP1 and UMAP2 by IDC
g <- umap_plot %>% ggplot(aes(x=V1, y=V2, color=disease)) +
  geom_point(size=1.5) + xlab("UMAP1") + ylab("UMAP2") +
  geom_text(label=colnames(se_cvid), nudge_y = 0.1,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("UMAP Plot - Entire Sample")

plot(g)
```

#### PCA (Coloring Treatment Groups)

```{r}
# fit PCA model to log cpm data
# note: transpose data to make it tidy
set.seed(1)
pca_out <- prcomp(t(assay(se_cvid,"log_counts_cpm")))

# define dataframe with PCA components and immune deficiency disease (IDC)
pca_plot <- as.data.frame(pca_out$x) # X, Y, ...
pca_plot$treatment <- as.factor(se_cvid$treatment) # color/group

# plot PC1 and PC2 by IDC
g <- pca_plot %>% ggplot(aes(x=PC1, y=PC2, color=treatment)) +
  geom_point(size=1.5) + xlab("PC 1") + ylab("PC 2") +
  geom_text(label=colnames(se_cvid), nudge_y = 5,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("PCA Plot - Entire Sample with Treatment Groups")

plot(g)
```

#### UMAP (Coloring Treatment Groups)

```{r}
# fit UMAP model to log cpm data
# note: transpose data to make it tidy
set.seed(1)
umap.defaults$n_neighbors=5
umap_out <- umap(t(assay(se_cvid,"log_counts_cpm")), config=umap.defaults)

# define dataframe with UMAP dimensions and IDC
umap_plot <- as.data.frame(umap_out$layout) # X, Y, ...
umap_plot$treatment <- as.factor(se_cvid$treatment) # color / group

# plot UMAP1 and UMAP2 by IDC
g <- umap_plot %>% ggplot(aes(x=V1, y=V2, color=treatment)) +
  geom_point(size=1.5) + xlab("UMAP1") + ylab("UMAP2") +
  geom_text(label=colnames(se_cvid), nudge_y = 0.1,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("UMAP Plot - Entire Sample with Treatment Groups")

plot(g)
```

As we can observe, the treatment variable overwhelms the dimension reduction, indicating we should run separate analysis for each treatment group.

# Group Analyses {.tabset}

## LPS Treatment Group

```{r LPS_vars, include=FALSE}
counts <- assay(se_cvid_LPS, "counts")
meta_data <- colData(se_cvid_LPS)
disease <- meta_data$disease
group <- meta_data$disease
```

### Visualization and Dimension reduction {.tabset}

#### PCA LPS Group

```{r}
# fit PCA model to log cpm data for LPS treatment group
# note: transpose data to make it tidy
set.seed(1)
pca_out <- prcomp(t(assay(se_cvid_LPS,"log_counts_cpm")))

# define dataframe with PCA components and immune deficiency disease (IDC)
pca_plot <- as.data.frame(pca_out$x) # X, Y, ...
pca_plot$disease <- as.factor(se_cvid_LPS$disease) # color/group

# plot PC1 and PC2 by IDC
g <- pca_plot %>% ggplot(aes(x=PC1, y=PC2, color=disease)) +
  geom_point(size=1.5) + xlab("PC 1") + ylab("PC 2") +
  geom_text(label=colnames(se_cvid_LPS), nudge_y = 5,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("PCA Plot - LPS group")

plot(g)
```

#### UMAP LPS Group

```{r}
# fit UMAP model to log cpm data
# note: transpose data to make it tidy
set.seed(1)
umap.defaults$n_neighbors=5
umap_out <- umap(t(assay(se_cvid_LPS,"log_counts_cpm")), config=umap.defaults)

# define dataframe with UMAP dimensions and IDC
umap_plot <- as.data.frame(umap_out$layout) # X, Y, ...
umap_plot$disease <- as.factor(se_cvid_LPS$disease) # color / group

# plot UMAP1 and UMAP2 by IDC
g <- umap_plot %>% ggplot(aes(x=V1, y=V2, color=disease)) +
  geom_point(size=1.5) + xlab("UMAP1") + ylab("UMAP2") +
  geom_text(label=colnames(se_cvid_LPS), nudge_y = 0.1,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("UMAP Plot - LPS group")

plot(g)
```

### Differential Expression {.tabset}

#### DESeq2 Analysis

```{r message=FALSE}
# format data into a DESeq2-friendly Data Structure
dds <- DESeqDataSetFromMatrix(countData=counts, colData=meta_data, design=~disease)
#colData is a data frame of demographic/phenotypic data

# filter for genes with significant expression  
# expression should be > 0, but > 100 should be better
dds<-dds[rowSums(counts(dds))>1,] #Gene Filtering

# fit negative binomial regression
dds<-DESeq(dds) #Performs estimation of size factors, dispersion, and negative binomial GLM f#itting

# extracting results, ordering them by adjusted p-value
res <- results(dds)[order(results(dds)[,6]),]
#res[1:10,]

# dislpay results for the top 1000 most significant genes
datatable(data.frame(res[1:1000,]))

# store the name of the top 250 most significant genes
deseq250 <- rownames(res)[1:250]
LPS.deseq.res <- data.frame(res)
```

#### Heatmap of DEGs (DESeq2)

```{r}
# Make a Heatmap of DEGs

# All together: extract the log cpm of the top 250 genes and store that in a matrix
# # get the name of the genes ordered by adjusted p-val
top_genes = rownames(results(dds)[order(results(dds)$padj),])[1:250]
# # convert log cpm to a matrix structure sorted by the top genes
mat = as.matrix(assay(se_cvid_LPS,"log_counts_cpm"))[top_genes,] 

# normalize data row-wise
mat = t(scale(t(mat)))

# store IDC into a single-column dataframe
df=data.frame(disease=colData(se_cvid_LPS)$disease) 

# create HeatMap with some annotations
ha_colors = list(disease=c("healthy"="Blue","Deficient"="Red"))

ha = HeatmapAnnotation(df = df, col = ha_colors)
Heatmap(mat,show_row_names=F,show_column_names = F, top_annotation=ha)

```

#### Limma Analysis

```{r}
# collapse data into healthy or not

# Store data in Limma-friendly Data Structure
dge <- DGEList(counts=counts, group=group) #From edgeR, Computes library size

# filter for genes with significant expression
counts<-counts[which(rowSums(cpm(counts))>1),] #Gene Filtering #(mask)
dge <- DGEList(counts=counts, group=group) #Re-compute library size #(filter)

# Trimmed-Mean Normalization of data
dge <- calcNormFactors(dge) #TMM normalization

# Create design matrix (intercept + indicator variable)

design<-model.matrix(~group)

# distribution normalizing transformation
v<-voom(dge, design) #voom transform to calculate weights to eliminate mean-variance #relationship
#use usual limma pipelines

# fit empirical bayes regression model
fit<-lmFit(v,design)
fit<-eBayes(fit)

# organize data by top ranked genes when predicting IDC (coef = col#2 = UCVID)
# topTable(fit, coef=2)

# display top 1000 most significant genes
datatable(topTable(fit, coef=2, number=1000))

# store the name of the top 250 most significant genes
limma250 <- rownames(topTable(fit, coef=2, number=250))
LPS.limma.res <- data.frame(topTable(fit, coef=2))
```

#### Heatmap of DEGs

```{r}
# Make a Heatmap of DEGs

# extract the log cpm of the top 250 genes and store that in a matrix
# get the name of the top 250 most significant genes
top_genes = rownames(topTable(fit, coef=2, number=250))
# convert log cpm to a matrix structure
mat = as.matrix(assay(se_cvid_LPS,"log_counts_cpm"))[top_genes,] 

# normalize data row-wise
mat = t(scale(t(mat)))

# store IDC into a single-column dataframe
df=data.frame(disease=colData(se_cvid_LPS)$disease) 

# create HeatMap with some annotations
ha = HeatmapAnnotation(df = df, col=ha_colors)
Heatmap(mat,show_row_names=F,show_column_names=F, top_annotation=ha)
```

### Pathway analysis {.tabset}

#### EnrichR analysis {.tabset}

##### DESeq2 Genes {.tabset}

```{r}
# based on the results of DEseq, cross-reference the gene with pathway databases to infer which pathways are the most significant

enriched <- enrichr(deseq250, dbs)
```

###### DESeq2 WikiPathways

```{r}
datatable(enriched$WikiPathways_2016)
```

###### DESeq2 KEGG

```{r}
datatable(enriched$KEGG_2016)
```

###### DESeq2 Panther

```{r}
datatable(enriched$Panther_2016)
```

###### DESeq2 Reactome

```{r}
datatable(enriched$Reactome_2016)
```

##### Limma Genes (CVID) {.tabset}

```{r}
enriched <- enrichr(limma250, dbs)
```

###### Limma WikiPathways

```{r}
datatable(enriched$WikiPathways_2016)
```

###### Limma KEGG

```{r}
datatable(enriched$KEGG_2016)
```

###### Limma Panther

```{r}
datatable(enriched$Panther_2016)
```

###### Limma Reactome

```{r}
datatable(enriched$Reactome_2016)
```

```{r LPS_out, include=FALSE}
LPS_deseq <- deseq250
LPS_limma <- limma250
```

## Healthy Control Group

```{r ctrl_vars, include=FALSE}
counts <- assay(se_cvid_ctrl, "counts")
meta_data <- colData(se_cvid_ctrl)
disease <- meta_data$disease
group <- meta_data$disease
```

### Visualization and Dimension reduction {.tabset}

#### PCA Healthy Group

```{r}
# fit PCA model to log cpm data for healthy control group
# note: transpose data to make it tidy
set.seed(1)
pca_out <- prcomp(t(assay(se_cvid_ctrl,"log_counts_cpm")))

# define dataframe with PCA components and immune deficiency disease (IDC)
pca_plot <- as.data.frame(pca_out$x) # X, Y, ...
pca_plot$disease <- as.factor(se_cvid_ctrl$disease) # color/group

# plot PC1 and PC2 by IDC
g <- pca_plot %>% ggplot(aes(x=PC1, y=PC2, color=disease)) +
  geom_point(size=1.5) + xlab("PC 1") + ylab("PC 2") +
  geom_text(label=colnames(se_cvid_ctrl), nudge_y = 5,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("PCA Plot - Control group")

plot(g)
```

#### UMAP Healthy Group

```{r}
# fit UMAP model to log cpm data
# note: transpose data to make it tidy
set.seed(1)
umap.defaults$n_neighbors=5
umap_out <- umap(t(assay(se_cvid_ctrl,"log_counts_cpm")), config=umap.defaults)

# define dataframe with UMAP dimensions and IDC
umap_plot <- as.data.frame(umap_out$layout) # X, Y, ...
umap_plot$disease <- as.factor(se_cvid_ctrl$disease) # color / group

# plot UMAP1 and UMAP2 by IDC
g <- umap_plot %>% ggplot(aes(x=V1, y=V2, color=disease)) +
  geom_point(size=1.5) + xlab("UMAP1") + ylab("UMAP2") +
  geom_text(label=colnames(se_cvid_ctrl), nudge_y = 0.1,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("UMAP of PBMC Unstimulated Samples")

#ggsave("umap_plot_pbmc_unstim_600dpi.jpeg", plot = g, width = 8, height = 6, units = "in", dpi = 600)

plot(g)
```

### Differential Expression {.tabset}

#### DESeq2 Analysis

```{r message=FALSE}
# format data into a DESeq2-friendly Data Structure
dds <- DESeqDataSetFromMatrix(countData=counts, colData=meta_data, design=~disease)
#colData is a data frame of demographic/phenotypic data

# filter for genes with significant expression  
# expression should be > 0, but > 100 should be better
dds<-dds[rowSums(counts(dds))>1,] #Gene Filtering

# fit negative binomial regression
dds<-DESeq(dds) #Performs estimation of size factors, dispersion, and negative binomial GLM f#itting

# extracting results, ordering them by adjusted p-value
res <- results(dds)[order(results(dds)[,6]),]
#res[1:10,]

# display results for the top 1000 most significant genes
datatable(data.frame(res[1:1000,]))

# store the name of the top 250 most significant genes
deseq250 <- rownames(res)[1:250]
PBMC.deseq.res <- data.frame(res)
```

#### Heatmap of DEGs (DESeq2)

```{r}
# Make a Heatmap of DEGs

# All together: extract the log cpm of the top 250 genes and store that in a matrix
# # get the name of the genes ordered by adjusted p-val
top_genes = rownames(results(dds)[order(results(dds)$padj),])[1:250]
# # convert log cpm to a matrix structure sorted by the top genes
mat = as.matrix(assay(se_cvid_ctrl,"log_counts_cpm"))[top_genes,] 

# normalize data row-wise
mat = t(scale(t(mat)))

# store IDC into a single-column dataframe
df=data.frame(disease=colData(se_cvid_ctrl)$disease) 

# create HeatMap with some annotations
ha_colors <-  list(disease=c("healthy"="Blue","Deficient"="Red"))

ha <- HeatmapAnnotation(df = df, col = ha_colors)
ht <- Heatmap(mat, show_row_names = FALSE, show_column_names = FALSE, 
              top_annotation = ha, name = "log2 Fold Change")

width_pixels <- 8 * 600
height_pixels <- 6 * 600

# save 600 DPI JPEG
jpeg("heatmap_pbmc_unstim_600dpi.jpeg", width = width_pixels, height = height_pixels, res = 600)
draw(ht)
dev.off()


```

#### Limma Analysis

```{r}
# collapse data into healthy or not

# Store data in Limma-friendly Data Structure
dge <- DGEList(counts=counts, group=group) #From edgeR, Computes library size

# filter for genes with significant expression
counts<-counts[which(rowSums(cpm(counts))>1),] #Gene Filtering #(mask)
dge <- DGEList(counts=counts, group=group) #Re-compute library size #(filter)

# Trimmed-Mean Normalization of data
dge <- calcNormFactors(dge) #TMM normalization

# Create design matrix (intercept + indicator variable)

design<-model.matrix(~group)

# distribution normalizing transformation
v<-voom(dge, design) #voom transform to calculate weights to eliminate mean-variance #relationship
#use usual limma pipelines

# fit empirical bayes regression model
fit<-lmFit(v,design)
fit<-eBayes(fit)

# organize data by top ranked genes when predicting IDC (coef = col#2 = UCVID)
# topTable(fit, coef=2)

# display top 1000 most significant genes
datatable(topTable(fit, coef=2, number=1000))

# store the name of the top 250 most significant genes
limma250 <- rownames(topTable(fit, coef=2, number=250))
PBMC.limma.res <- data.frame(topTable(fit, coef=2))
```

#### Heatmap of DEGs

```{r}
# Make a Heatmap of DEGs

# extract the log cpm of the top 250 genes and store that in a matrix
# get the name of the top 250 most significant genes
top_genes = rownames(topTable(fit, coef=2, number=250))
# convert log cpm to a matrix structure
mat = as.matrix(assay(se_cvid_ctrl,"log_counts_cpm"))[top_genes,] 

# normalize data row-wise
mat = t(scale(t(mat)))

# store IDC into a single-column dataframe
df=data.frame(disease=colData(se_cvid_ctrl)$disease) 

# create HeatMap with some annotations
ha = HeatmapAnnotation(df = df, col=ha_colors)
Heatmap(mat,show_row_names=F,show_column_names=F, top_annotation=ha)
```

### Pathway analysis {.tabset}

#### EnrichR analysis {.tabset}

##### DESeq2 Genes {.tabset}

```{r}
# based on the results of DEseq, cross-reference the gene with pathway databases to infer which pathways are the most significant

enriched <- enrichr(deseq250, dbs)
```

###### DESeq2 WikiPathways

```{r}
datatable(enriched$WikiPathways_2016)
```

###### DESeq2 KEGG

```{r}
datatable(enriched$KEGG_2016)
```

###### DESeq2 Panther

```{r}
datatable(enriched$Panther_2016)
```

###### DESeq2 Reactome

```{r}
datatable(enriched$Reactome_2016)
```

##### Limma Genes (CVID) {.tabset}

```{r}
enriched <- enrichr(limma250, dbs)
```

###### Limma WikiPathways

```{r}
datatable(enriched$WikiPathways_2016)
```

###### Limma KEGG

```{r}
datatable(enriched$KEGG_2016)
```

###### Limma Panther

```{r}
datatable(enriched$Panther_2016)
```

###### Limma Reactome

```{r}
datatable(enriched$Reactome_2016)
```

```{r ctrl_out, include=FALSE}
ctrl_deseq <- deseq250
ctrl_limma <- limma250

```


### Creating CSV for Comparative Analysis
```{r}
#combine result lists into dataframe

top.250.results.df <- data.frame(ctrl_deseq=ctrl_deseq,
                           ctrl_limma=ctrl_limma,
                           LPS_deseq=LPS_deseq,
                           LPS_limma=LPS_limma) 

# obtain project directory
project_root <- rprojroot::find_root(has_dir("Meta-omic-Analysis-Capstone"))
# obtain comparison analysis directory

relative_path <- file.path("Meta-omic-Analysis-Capstone/Deliverable 6")

# file output paths
output.path.1 <- fs::path(project_root,
                         relative_path, "data",
                         "PBMC_CVID_top250_results.csv")
output.path.2 <- fs::path(project_root,
                         relative_path, "data",
                         "PBMC_CVID_all_results.csv")

# write .csv file into Deliverable 6 folder
write.csv(x=top.250.results.df, file=output.path.1, row.names = T)
write.csv(x=PBMC.deseq.res, file=output.path.2, row.names = T)

```
