---
title: "Analysis of Immune Deficiency RNA-Seq Data"
author: W. Evan Johnson, Ph.D., Ian Mc Farlane, Bram Stults
date: "`r Sys.Date()`"
header-includes:
   - \usepackage{amsmath}
output:
  html_document:
    code_folding: show
    toc: true
    toc_float: true
    theme: "flatly"
editor_options: 
  chunk_output_type: console
tables: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

# load required libraries
library(Rsubread)
library(Rsamtools)
library(SummarizedExperiment)
library(edgeR)
library(DESeq2)
library(ComplexHeatmap)
library(TBSignatureProfiler)
library(umap)
library(tidyverse)
library(DT)
library(GSVA)
library(enrichR)
library(biomaRt)
library(rprojroot)
library(fs)
library(dplyr)
library(comprehenr)

# define pathway libraries
dbs <- c("WikiPathways_2016","KEGG_2016","Panther_2016","Reactome_2016")
```

### Read in dataset

```{r}
# read in and format patient data for both blood and nasal samples
meta_data <- read.csv("data/full_metadata.csv", row.names = 1)
meta_data$sample_name_rnacounts_file <- toupper(meta_data$sample_name_rnacounts_file)
meta_data$disease <- meta_data$status

# read expression counts
blood.counts.df <- read.csv("data/bloodcounts.csv", row.names = 1)
nasal.counts.df <- read.csv("data/nasalcounts.csv", row.names = 1)

# filter patient data without a match on expression count dataframes
relevant.rows <- meta_data$sample_name_rnacounts_file %in% c(names(blood.counts.df), names(nasal.counts.df)) 
meta_data <- meta_data[relevant.rows,]
meta_data$status <- factor(meta_data$status,levels = c("control","case"))

# split patient data into blood and nasal samples
blood.meta_data <- meta_data[meta_data$type == "blood",]
nasal.meta_data <- meta_data[meta_data$type == "nasal",]

# identify patients that performed both trials
samples_of_interest <- intersect(blood.meta_data$sample_id, nasal.meta_data$sample_id)

# blood: filter for the patients that also participated in nasal sampling
blood.mask <- blood.meta_data$sample_id %in% samples_of_interest
blood.meta_data <- blood.meta_data[blood.mask,]
columns <- blood.meta_data$sample_name_rnacounts_file
blood.counts.df <- blood.counts.df[,columns]

# nasal: filter for the patients that also participated in blood sampling
nasal.mask <- nasal.meta_data$sample_id %in% samples_of_interest
nasal.meta_data <- nasal.meta_data[nasal.mask,]
columns <- nasal.meta_data$sample_name_rnacounts_file
nasal.counts.df <- nasal.counts.df[,columns]


# aggregate patient and expression data into a single object for each set of samples
se.blood <- SummarizedExperiment(assays=list(counts=blood.counts.df), colData=blood.meta_data)
se.nasal <- SummarizedExperiment(assays=list(counts=nasal.counts.df), colData=nasal.meta_data)

# add log*cpm assays
se.blood <- mkAssay(se.blood, log = TRUE, counts_to_CPM = TRUE)
se.nasal <- mkAssay(se.nasal, log = TRUE, counts_to_CPM = TRUE)

# display all assays, blood only
assays(se.blood)

# display all assays, nasal only
assays(se.nasal)
```



# Group Analyses {.tabset}

## Blood Treatment Group

```{r blood_vars, include=FALSE}
counts <- assay(se.blood, "counts")
meta_data <- colData(se.blood)
disease <- meta_data$status
group <- meta_data$status
x <- meta_data$sample_id
labels <- sapply(strsplit(x, "-"), function(parts) {
  paste0(as.numeric(parts[2]), "-", as.numeric(parts[3]))
})
```

### Visualization and Dimension reduction {.tabset}

#### PCA Blood Treatment

```{r}
# fit PCA model to log cpm data for blood treatment group
# note: transpose data to make it tidy
set.seed(1)
pca_out <- prcomp(t(assay(se.blood,"log_counts_cpm")))

# define dataframe with PCA components and immune deficiency disease (IDC)
pca_plot <- as.data.frame(pca_out$x) # X, Y, ...
pca_plot$disease <- as.factor(disease) # color/group

# plot PC1 and PC2 by IDC
g <- pca_plot %>% ggplot(aes(x=PC1, y=PC2, color=disease)) +
  geom_point(size=1.5) + xlab("PC 1") + ylab("PC 2") +
  geom_text(label=labels, nudge_y = 5,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("PCA Plot - Blood Treatment")

plot(g)
```

#### UMAP Blood Treatment

```{r}
# fit UMAP model to log cpm data
# note: transpose data to make it tidy
set.seed(1)
umap.defaults$n_neighbors=5
umap_out <- umap(t(assay(se.blood,"log_counts_cpm")), config=umap.defaults)

# define dataframe with UMAP dimensions and IDC
umap_plot <- as.data.frame(umap_out$layout) # X, Y, ...
umap_plot$disease <- as.factor(disease) # color / group

# plot UMAP1 and UMAP2 by IDC
g <- umap_plot %>% ggplot(aes(x=V1, y=V2, color=disease)) +
  geom_point(size=1.5) + xlab("UMAP1") + ylab("UMAP2") +
  geom_text(label=labels, nudge_y = 0.1,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("UMAP Plot - Blood Treatment")

plot(g)
```

### Differential Expression {.tabset}

#### DESeq2 Analysis

```{r message=FALSE}
# format data into a DESeq2-friendly Data Structure
dds <- DESeqDataSetFromMatrix(countData=counts, colData=meta_data, design=~disease)
#colData is a data frame of demographic/phenotypic data

# filter for genes with significant expression  
# expression should be > 0, but > 100 should be better
dds<-dds[rowSums(counts(dds))>1,] #Gene Filtering

# fit negative binomial regression
dds<-DESeq(dds) #Performs estimation of size factors, dispersion, and negative binomial GLM f#itting

# extracting results, ordering them by adjusted p-value
res <- results(dds)[order(results(dds)[,6]),]
#res[1:10,]

# display results for the top 1000 most significant genes
datatable(data.frame(res[1:1000,]))

# store the name of the top 250 most significant genes
deseq250 <- rownames(res)[1:250]
blood.res <- data.frame(res)
```

#### Heatmap of DEGs (DESeq2)

```{r}
# Make a Heatmap of DEGs

# All together: extract the log cpm of the top 250 genes and store that in a matrix
# # get the name of the genes ordered by adjusted p-val
top_genes = rownames(results(dds)[order(results(dds)$padj),])[1:250]
# # convert log cpm to a matrix structure sorted by the top genes
mat = as.matrix(assay(se.blood,"log_counts_cpm"))[top_genes,] 

# normalize data row-wise
mat = t(scale(t(mat)))

# store IDC into a single-column dataframe
df=data.frame(disease=colData(se.blood)$disease) 

# create HeatMap with some annotations
ha_colors = list(disease=c("control"="Blue","case"="Red"))

ha = HeatmapAnnotation(df = df, col = ha_colors)
Heatmap(mat,show_row_names=F,show_column_names = F, top_annotation=ha)

```

#### Limma Analysis

```{r}
# collapse data into healthy or not

# Store data in Limma-friendly Data Structure
dge <- DGEList(counts=counts, group=group) #From edgeR, Computes library size

# filter for genes with significant expression
counts<-counts[which(rowSums(cpm(counts))>1),] #Gene Filtering #(mask)
dge <- DGEList(counts=counts, group=group) #Re-compute library size #(filter)

# Trimmed-Mean Normalization of data
dge <- calcNormFactors(dge) #TMM normalization

# Create design matrix (intercept + indicator variable)

design<-model.matrix(~group)

# distribution normalizing transformation
v<-voom(dge, design) #voom transform to calculate weights to eliminate mean-variance #relationship
#use usual limma pipelines

# fit empirical bayes regression model
fit<-lmFit(v,design)
fit<-eBayes(fit)

# organize data by top ranked genes when predicting IDC (coef = col#2 = UCVID)
# topTable(fit, coef=2)

# display top 1000 most significant genes
datatable(topTable(fit, coef=2, number=1000))

# store the name of the top 250 most significant genes
limma250 <- rownames(topTable(fit, coef=2, number=250))
```

#### Heatmap of DEGs

```{r}
# Make a Heatmap of DEGs

# extract the log cpm of the top 250 genes and store that in a matrix
# get the name of the top 250 most significant genes
top_genes = rownames(topTable(fit, coef=2, number=250))
# convert log cpm to a matrix structure
mat = as.matrix(assay(se.blood,"log_counts_cpm"))[top_genes,] 

# normalize data row-wise
mat = t(scale(t(mat)))

# store IDC into a single-column dataframe
df=data.frame(disease=colData(se.blood)$disease) 

# create HeatMap with some annotations
ha = HeatmapAnnotation(df = df, col=ha_colors)
Heatmap(mat,show_row_names=F,show_column_names=F, top_annotation=ha)
```

### Pathway analysis {.tabset}

#### EnrichR analysis {.tabset}

##### DESeq2 Genes {.tabset}

```{r}
# based on the results of DEseq, cross-reference the gene with pathway databases to infer which pathways are the most significant

enriched <- enrichr(deseq250, dbs)
```

###### DESeq2 WikiPathways

```{r}
datatable(enriched$WikiPathways_2016)
```

###### DESeq2 KEGG

```{r}
datatable(enriched$KEGG_2016)
```

###### DESeq2 Panther

```{r}
datatable(enriched$Panther_2016)
```

###### DESeq2 Reactome

```{r}
datatable(enriched$Reactome_2016)
```

##### Limma Genes (CVID) {.tabset}

```{r}
enriched <- enrichr(limma250, dbs)
```

###### Limma WikiPathways

```{r}
datatable(enriched$WikiPathways_2016)
```

###### Limma KEGG

```{r}
datatable(enriched$KEGG_2016)
```

###### Limma Panther

```{r}
datatable(enriched$Panther_2016)
```

###### Limma Reactome

```{r}
datatable(enriched$Reactome_2016)
```

```{r LPS_out, include=FALSE}
blood.deseq <- deseq250
blood.limma <- limma250
```

## Nasal Treatment Group

```{r nasal_vars, include=FALSE}
counts <- assay(se.nasal, "counts")
meta_data <- colData(se.nasal)
disease <- meta_data$status
group <- meta_data$status
x <- meta_data$sample_id
labels <- sapply(strsplit(x, "-"), function(parts) {
  paste0(as.numeric(parts[2]), "-", as.numeric(parts[3]))
})
```

### Visualization and Dimension reduction {.tabset}

#### PCA Nasal Treatment

```{r}
# fit PCA model to log cpm data for healthy control group
# note: transpose data to make it tidy
set.seed(1)
pca_out <- prcomp(t(assay(se.nasal,"log_counts_cpm")))

# define dataframe with PCA components and immune deficiency disease (IDC)
pca_plot <- as.data.frame(pca_out$x) # X, Y, ...
pca_plot$disease <- as.factor(disease) # color/group

# plot PC1 and PC2 by IDC
g <- pca_plot %>% ggplot(aes(x=PC1, y=PC2, color=disease)) +
  geom_point(size=1.5) + xlab("PC 1") + ylab("PC 2") +
  geom_text(label=labels, nudge_y = 5,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("PCA Plot - Nasal Treatment")

plot(g)
```

#### UMAP Nasal Treatment

```{r}
# fit UMAP model to log cpm data
# note: transpose data to make it tidy
set.seed(1)
umap.defaults$n_neighbors=5
umap_out <- umap(t(assay(se.nasal,"log_counts_cpm")), config=umap.defaults)

# define dataframe with UMAP dimensions and IDC
umap_plot <- as.data.frame(umap_out$layout) # X, Y, ...
umap_plot$disease <- as.factor(se.nasal$disease) # color / group

# plot UMAP1 and UMAP2 by IDC
g <- umap_plot %>% ggplot(aes(x=V1, y=V2, color=disease)) +
  geom_point(size=1.5) + xlab("UMAP1") + ylab("UMAP2") +
  geom_text(label=labels, nudge_y = 0.1,) +
  theme(plot.title = element_text(hjust = 0.5)) +
  ggtitle("UMAP Plot - Nasal Treatment")

plot(g)
```

### Differential Expression {.tabset}

#### DESeq2 Analysis

```{r message=FALSE}
# format data into a DESeq2-friendly Data Structure
dds <- DESeqDataSetFromMatrix(countData=counts, colData=meta_data, design=~disease)
#colData is a data frame of demographic/phenotypic data

# filter for genes with significant expression  
# expression should be > 0, but > 100 should be better
dds<-dds[rowSums(counts(dds))>1,] #Gene Filtering

# fit negative binomial regression
dds<-DESeq(dds) #Performs estimation of size factors, dispersion, and negative binomial GLM f#itting

# extracting results, ordering them by adjusted p-value
res <- results(dds)[order(results(dds)[,6]),]
#res[1:10,]

# display results for the top 1000 most significant genes
datatable(data.frame(res[1:1000,]))

# store the name of the top 250 most significant genes
deseq250 <- rownames(res)[1:250]
nasal.res <- data.frame(res)
```

#### Heatmap of DEGs (DESeq2)

```{r}
# Make a Heatmap of DEGs

# All together: extract the log cpm of the top 250 genes and store that in a matrix
# # get the name of the genes ordered by adjusted p-val
top_genes = rownames(results(dds)[order(results(dds)$padj),])[1:250]
# # convert log cpm to a matrix structure sorted by the top genes
mat = as.matrix(assay(se.nasal,"log_counts_cpm"))[top_genes,] 

# normalize data row-wise
mat = t(scale(t(mat)))

# store IDC into a single-column dataframe
df=data.frame(disease=colData(se.nasal)$disease) 

ha = HeatmapAnnotation(df = df, col = ha_colors)
Heatmap(mat,show_row_names=F,show_column_names = F, top_annotation=ha)

```

#### Limma Analysis

```{r}
# collapse data into healthy or not

# Store data in Limma-friendly Data Structure
dge <- DGEList(counts=counts, group=group) #From edgeR, Computes library size

# filter for genes with significant expression
counts<-counts[which(rowSums(cpm(counts))>1),] #Gene Filtering #(mask)
dge <- DGEList(counts=counts, group=group) #Re-compute library size #(filter)

# Trimmed-Mean Normalization of data
dge <- calcNormFactors(dge) #TMM normalization

# Create design matrix (intercept + indicator variable)

design<-model.matrix(~group)

# distribution normalizing transformation
v<-voom(dge, design) #voom transform to calculate weights to eliminate mean-variance #relationship
#use usual limma pipelines

# fit empirical bayes regression model
fit<-lmFit(v,design)
fit<-eBayes(fit)

# organize data by top ranked genes when predicting IDC (coef = col#2 = UCVID)
# topTable(fit, coef=2)

# display top 1000 most significant genes
datatable(topTable(fit, coef=2, number=1000))

# store the name of the top 250 most significant genes
limma250 <- rownames(topTable(fit, coef=2, number=250))
```

#### Heatmap of DEGs

```{r}
# Make a Heatmap of DEGs

# extract the log cpm of the top 250 genes and store that in a matrix
# get the name of the top 250 most significant genes
top_genes = rownames(topTable(fit, coef=2, number=250))
# convert log cpm to a matrix structure
mat = as.matrix(assay(se.nasal,"log_counts_cpm"))[top_genes,] 

# normalize data row-wise
mat = t(scale(t(mat)))

# store IDC into a single-column dataframe
df=data.frame(disease=colData(se.nasal)$disease) 

# create HeatMap with some annotations
ha = HeatmapAnnotation(df = df, col=ha_colors)
Heatmap(mat,show_row_names=F,show_column_names=F, top_annotation=ha)
```

### Pathway analysis {.tabset}

#### EnrichR analysis {.tabset}

##### DESeq2 Genes {.tabset}

```{r}
# based on the results of DEseq, cross-reference the gene with pathway databases to infer which pathways are the most significant

enriched <- enrichr(deseq250, dbs)
```

###### DESeq2 WikiPathways

```{r}
datatable(enriched$WikiPathways_2016)
```

###### DESeq2 KEGG

```{r}
datatable(enriched$KEGG_2016)
```

###### DESeq2 Panther

```{r}
datatable(enriched$Panther_2016)
```

###### DESeq2 Reactome

```{r}
datatable(enriched$Reactome_2016)
```

##### Limma Genes (CVID) {.tabset}

```{r}
enriched <- enrichr(limma250, dbs)
```

###### Limma WikiPathways

```{r}
datatable(enriched$WikiPathways_2016)
```

###### Limma KEGG

```{r}
datatable(enriched$KEGG_2016)
```

###### Limma Panther

```{r}
datatable(enriched$Panther_2016)
```

###### Limma Reactome

```{r}
datatable(enriched$Reactome_2016)
```

```{r ctrl_out, include=FALSE}
nasal.deseq <- deseq250
nasal.limma <- limma250

```



# Comparative analysis

## Overview
### Common Significant genes

```{r}
(common_deseq <- intersect(blood.deseq, nasal.deseq))
(common_limma <- intersect(blood.limma, nasal.limma))
```

### Similarity Coefficients for Limma, DESeq analyses

```{r}
print("DESeq similarity coefficient")
length(common_deseq) / length(union(blood.deseq, nasal.deseq))

print("LIMMA similarity coefficient")
length(common_limma) / length(union(blood.limma, nasal.limma))
```


## Combined DEGs (DeSeq)
```{r}
common_genes <- common_deseq
significance_threshold <- 0.3

combined_results <- data.frame(gene = common_genes) %>%
    mutate(
        blood.padj = blood.res[common_genes,]$padj,
        nasal.padj = nasal.res[common_genes,]$padj,
        combined_chisq = -2*(log(blood.padj) + log(nasal.padj)),
        blood.effect = blood.res[common_genes,]$log2FoldChange,
        nasal.effect = nasal.res[common_genes,]$log2FoldChange,
    ) %>% 
    transmute(
        gene = gene, blood.padj = blood.padj, nasal.padj = nasal.padj,
        combined_p = 1 - pchisq(combined_chisq, 4),
        blood.effect = ifelse(blood.effect > 0, "Upregulated", "Downregulated"),
        nasal.effect = ifelse(nasal.effect > 0, "Upregulated", "Downregulated"),
        reg_agreement = ifelse(blood.effect == nasal.effect, "Same", "Opposite"),
    ) %>% 
    mutate(across(where(is.character), as.factor))%>% 
    filter(
        !is.na(combined_p),
        blood.padj < significance_threshold,
        nasal.padj < significance_threshold,
    ) %>%
    arrange(combined_p) %>% 
    slice_head(n = 250)

row.names(combined_results) <- combined_results$gene
combined_results$gene <- NULL
combined_results$nasal_padj <- NULL
combined_results$ctrl_padj <- NULL


datatable(combined_results)
```

## Monte Carlo Simulation of Jaccard Index

```{r}

# function for commonality coefficient 
commonality_coefficient <- function(sample1, sample2) {
  intersection <- intersect(rownames(sample1), rownames(sample2))
  union <- union(rownames(sample1), rownames(sample2))
  return(length(intersection) / length(union))
}

# number of simulations
num_sims <- 100000

# number of genes to sample
sample_size <- 250

# result vector
common_coeffs <- numeric(num_sims)

# Loop
for (i in 1:num_sims) {
  
  # randomly selecting 250 row indices from full blood.res results
  blood.indices <- sample(nrow(blood.res), 
                          size = sample_size, 
                          replace = F) #
  blood.sample <- blood.res[blood.indices, ]
  
  # randomly selecting 250 row indices from full results nasal.res
  nasal.indices <- sample(nrow(nasal.res), 
                      size = sample_size, 
                      replace = F) # 
  nasal.sample <- nasal.res[nasal.indices, ]

  # calculating and storing commonality coefficient for samples
  coefficient <- commonality_coefficient(blood.sample, nasal.sample)
  common_coeffs[i] <- coefficient
}

mean_commonality <- mean(common_coeffs)
sd_commonality <- sd(common_coeffs)

cat("Mean Commonality Coefficient:", mean_commonality, "\n")
cat("Standard deviation:", sd_commonality, "\n")

hist(common_coeffs, 
     main = "Distribution of Commonality Coefficients", 
     xlab = "Commonality Coefficient")

#Probability of obtaining Jaccard index greater than obtained Deseq coefficient
mean(common_coeffs >= 0.02669405) #mean of True/False logical vector T/F --> 1/0



```
If the 250 genes in a list were completely random, there is a 0.0004% chance that the resulting commonality coefficients of those lists would be as large as or larger than 0.02669405 obtained here.

## Pathways
```{r}

enriched <- enrichr(rownames(combined_results), dbs)


combined_df <- bind_rows(
      enriched$WikiPathways_2016 %>% mutate(database = "WikiPathways"),
      enriched$KEGG_2016 %>% mutate(database = "KEGG"),
      enriched$Panther_2016 %>% mutate(database = "Panther"),
      enriched$Reactome_2016 %>% mutate(database = "Reactome")
    ) %>%
    arrange(Adjusted.P.value)

combined_df$Adjusted.P.value <- NULL
combined_df$Old.Adjusted.P.value <- NULL

# Display the combined data frame
datatable(combined_df)
```

=======

### Creating CSV for Comparative Analysis
```{r include = FALSE, eval = FALSE, echo = FALSE}
#combine result lists into dataframe

tb_results_df <- data.frame(blood.deseq=blood.deseq,
                           blood.limma=blood.limma,
                           nasal.deseq=nasal.deseq,
                           nasal.limma=nasal.limma) 

# obtain project directory
project_root <- rprojroot::find_root(has_dir("Meta-omic-Analysis-Capstone"))
# obtain comparison analysis directory
relative_path <- file.path("Meta-omic-Analysis-Capstone/Deliverable 5")
# file's output path
output_path <- fs::path(project_root,
                         relative_path, "data",
                         "TB_results_top250.csv")
# write .csv file into Deliverable 5 folder
write_csv(x=tb_results_df, file=output_path)
```

### Permutation Tests

#### Using the whole dataset {.tabset}


##### Assuming independent regulation
```{r permutation_test, collapse=TRUE}
# simulation parameters
N <- 10000
empirical_agreement <- mean(combined_results$reg_agreement == "Same")

blood.effect <- ifelse(blood.res$log2FoldChange > 0, "Up Regulated", "Down Regulated")
nasal.effect <- ifelse(nasal.res$log2FoldChange > 0, "Up Regulated", "Down Regulated")

blood.dist <- table(blood.effect)
nasal.dist <- table(nasal.effect)

blood.sampler <- \(n) sample(names(blood.dist), size=n, replace=TRUE, prob=blood.dist)
nasal.sampler <- \(n) sample(names(nasal.dist), size=n, replace=TRUE, prob=nasal.dist)

blood.genes <- row.names(blood.res)
nasal.genes <- row.names(blood.res)

common.genes <- intersect(blood.genes, nasal.genes)
common.count <- length(common.genes)

permutations <- to_vec(for(i in 1:N){
    
    # extract the regulation direction for the most significant genes
    blood.effect <- blood.sampler(common.count)
    nasal.effect <- nasal.sampler(common.count)
    
    # get an agreement coefficient (# agreed / total)
    mean(blood.effect == nasal.effect)
})

# display p-val and counts
mean(permutations  < empirical_agreement)
table(permutations  < empirical_agreement)
hist(permutations)
median(permutations)
```

##### Assuming dependent regulation
```{r collapse=TRUE}
joint.res <- merge(blood.res, nasal.res, by.x = 0, by.y = 0)
joint.effect <- ifelse(joint.res[c("log2FoldChange.x", "log2FoldChange.y")] > 0, "Up Regulated", "Down Regulated")

joint.dist <- as.data.frame(table(joint.effect[,1], joint.effect[,2]))
joint.sampler <-  \(n) joint.dist[sample(1:nrow(joint.dist), size=n, replace=TRUE, prob=joint.dist$Freq), ]

permutations <- to_vec(for(i in 1:N){
    
    # extract the regulation direction for the most significant genes
    joint.effect <- joint.sampler(common.count)
    blood.effect <- joint.effect$Var1
    nasal.effect <- joint.effect$Var2
    
    # get an agreement coefficient (# agreed / total)
    mean(blood.effect == nasal.effect)
})

# display p-val and counts
mean(permutations  < empirical_agreement)
table(permutations  < empirical_agreement)
hist(permutations)
median(permutations)
```

#### Selecting the top 250 genes {.tabset}

##### Assuming independent regulation
```{r collapse=TRUE}
N <- 2000
permutations <- to_vec(for(i in 1:N){
    
    blood.shuffle <- data.frame(
        padj = blood.res$padj,
        effect = blood.sampler(nrow(blood.res))
    )
    
    nasal.shuffle <- data.frame(
        padj = nasal.res$padj,
        effect = nasal.sampler(nrow(nasal.res))
    )
        
    row.names(blood.shuffle) <- sample(row.names(blood.res), nrow(blood.res))
    row.names(nasal.shuffle) <- sample(row.names(nasal.res), nrow(nasal.res))
    
    # extract the top 250 genes for each experiment
    blood.250 <- row.names(head(arrange(blood.shuffle, padj), 250))
    nasal.250 <- row.names(head(arrange(nasal.shuffle, padj), 250))
    
    # identify genes significant in both experiments
    common.genes <- intersect(blood.250, nasal.250)
   
    # extract the regulation direction for the most significant genes
    blood.effect <- blood.shuffle[common.genes,]$effect
    nasal.effect <- nasal.shuffle[common.genes,]$effect
    
    # get an agreement coefficient (# agreed / total)
    if (sum(blood.effect == nasal.effect) == 0) {
        agreement <- 1
    } else {
        agreement <- mean(blood.effect == nasal.effect)
    }
})

# display p-val and counts
mean(permutations  < empirical_agreement)
table(permutations  < empirical_agreement)
hist(permutations)
median(permutations)
```

##### Assuming dependent regulation
```{r collapse=TRUE}
blood.genes <- row.names(blood.res)
nasal.genes <- row.names(blood.res)

common.genes.static <- intersect(blood.genes, nasal.genes)
common.count <- length(common.genes.static)

blood.res.common <- blood.res[common.genes,]
nasal.res.common <- nasal.res[common.genes,]


permutations <- to_vec(for(i in 1:N){
    
    joint.effect <- joint.sampler(common.count)
    
    blood.shuffle <- data.frame(
        padj = blood.res.common$padj,
        effect = joint.effect$Var1
    )
    
    nasal.shuffle <- data.frame(
        padj = nasal.res.common$padj,
        effect = joint.effect$Var2
    )
        
    row.names(blood.shuffle) <- sample(common.genes.static, common.count)
    row.names(nasal.shuffle) <- sample(common.genes.static, common.count)
    
    # extract the top 250 genes for each experiment
    blood.250 <- row.names(head(arrange(blood.shuffle, padj), 250))
    nasal.250 <- row.names(head(arrange(nasal.shuffle, padj), 250))
    
    # identify genes significant in both experiments
    common.genes <- intersect(blood.250, nasal.250)
   
    # extract the regulation direction for the most significant genes
    blood.effect <- blood.shuffle[common.genes,]$effect
    nasal.effect <- nasal.shuffle[common.genes,]$effect
    
    # get an agreement coefficient (# agreed / total)
    if (sum(blood.effect == nasal.effect) == 0) {
        agreement <- 1
    } else {
        agreement <- mean(blood.effect == nasal.effect)
    }
    
})

# display p-val and counts
mean(permutations  < empirical_agreement)
table(permutations  < empirical_agreement)
hist(permutations)
median(permutations)
```


##### Assuming dependent regulation and regulation tied to p-value
```{r collapse=TRUE}

# auxiliary dataframes
blood.shuffle <- data.frame(
    padj = blood.res$padj,                
    effect = ifelse(blood.res$log2FoldChange > 0, "Up Regulated", "Down Regulated")
)

nasal.shuffle <- data.frame(
    padj = nasal.res$padj,                
    effect = ifelse(nasal.res$log2FoldChange > 0, "Up Regulated", "Down Regulated")
)
   
permutations <- to_vec(for(i in 1:N){
   
    # shuffle gene names
    row.names(blood.shuffle) <- sample(row.names(blood.res), nrow(blood.res))
    row.names(nasal.shuffle) <- sample(row.names(nasal.res), nrow(nasal.res))
   
    # extract the top 250 genes for each experiment
    blood.250 <- row.names(head(arrange(blood.shuffle, padj), 250))
    nasal.250 <- row.names(head(arrange(nasal.shuffle, padj), 250))
   
    # identify genes significant in both experiments
    common.genes <- intersect(blood.250, nasal.250)
   
    # extract the regulation direction for the most significant genes
    blood.effect <- blood.shuffle[common.genes,]$effect
    nasal.effect <- nasal.shuffle[common.genes,]$effect
   
    # get an agreement coefficient (# agreed / total)
    if (sum(blood.effect == nasal.effect) == 0) {
        agreement <- 1
    } else {
        agreement <- mean(blood.effect == nasal.effect)
    }
   
    # report if agreement coefficient is more extreme
    agreement
   
})

# display p-val and counts
mean(permutations  < empirical_agreement)
table(permutations  < empirical_agreement)
hist(permutations)
median(permutations)
``` 


